cassandra_offset_storage.go

```diff
@@ -0,0 +1,65 @@
+package go_kafka_client
+
+import (
+    cql "github.com/gocql/gocql"
+    "sync"
+    "fmt"
+    "strings"
+)
+
+type CassandraOffsetStorage struct {
+    cluster *cql.ClusterConfig
+    connection *cql.Session
+    lock sync.Mutex
+}
+
+func NewCassandraOffsetStorage(host string, keyspace string) *CassandraOffsetStorage {
+    cluster := cql.NewCluster(host)
+    cluster.Keyspace = keyspace
+
+    return &CassandraOffsetStorage{cluster: cluster,}
+}
+
+// Gets the offset for a given group, topic and partition.
+// May return an error if fails to retrieve the offset.
+func (this *CassandraOffsetStorage) GetOffset (group string, topic string, partition int32) (int64, error) {
+    this.ensureConnection(group)
+    response := &OffsetResponse{}
+    data := this.connection.Query(fmt.Sprintf("SELECT * FROM %s WHERE topic = ? AND partition = ? LIMIT 1", strings.Replace(group, "-", "", -1)), topic, partition).Iter()
+    data.Scan(&response.Topic, &response.Partition, &response.LastOffset, &response.TraceData)
+    if err := data.Close(); err != nil {
+        return InvalidOffset, err
+    }
+
+    return response.LastOffset, nil
+}
+
+// Commits the given offset for a given group, topic and partition.
+// May return an error if fails to commit the offset.
+func (this *CassandraOffsetStorage) CommitOffset (group string, topic string, partition int32, offset int64) error {
+    this.ensureConnection(group)
+    return this.connection.Query(fmt.Sprintf("UPDATE %s SET last_offset = ? WHERE topic = ? AND partition = ?", strings.Replace(group, "-", "", -1)), offset, topic, partition).Exec()
```

Comment:
we should store more information in here than just the offset, we should have the timestamp (now()) of when it happened. 

Eventually (not sure we can do this just yet) we also want to store the trace information for the client at that time. In many cases we can't or won't want to add the "timings" to the message because it artificially inflates traffic. It "might" make sense to store that here so we can see over time for each offset being saved what was going on. I would prefer that "trace saved" operation being for an existing operation so as to not add more moving parts.